ROX-25572: prefetch dependencies for operator-bundle on Konflux (#13209)
stackrox/stackrox: operator/bundle_helpers/README.md

# Bundle Helpers

> ## Dependency Management

Reviewer: ```suggestion
## Python Dependency Management
```

Author: https://github.com/stackrox/stackrox/pull/13209/commits/ab8529e746514f797378710cc56b0aee6e660bb7

done

> 

Author: https://github.com/stackrox/stackrox/pull/13209/commits/ab8529e746514f797378710cc56b0aee6e660bb7

done

Dependencies are managed with Poetry.
This tool can generate the full dependency tree for Python dependencies.
Install it from [here](https://python-poetry.org/docs/#installation).
You also need Python==3.6 installed as a prerequisite to keep GHA builds working.
This is going to be solved by [ROX-26860](https://issues.redhat.com/browse/ROX-26860).

In this directory, run

> * to add a new dependency: `poetry add PyYAML==6.0`

Reviewer: It deserves a mention that for as long as we have Downstream builds (hopefully not for long but still), all dependencies _actually_ used by scripts must be provided OOTB in CPaaS `render_templates` context.
CPaaS does not respect `requirements.txt` or anything and we can only use things that image maintainers included.
PyYAML is one such thing provided there and we can safely use it.

I did some basic thing to dump the python's version and list packages available in CPaaS but it was a while ago and things have changed.

I mention this because when you write `to add a new dependency`, it should be noted that we can't simply add any dependencies (and select any versions) we want while CPaaS is there.

Could you please somehow transform this in the note next to this list item?

Author: > CPaaS does not respect requirements.txt or anything and we can only use things that image maintainers included.

It used to do that before https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/commit/743a32a48bf967522aba7db54f91bc918f5a1cd9#609875223a8e1e7c7bc513ef01085c5fdbe00401_69_64, so we could install more dependencies here, right? (this is a very theoretical point, as we plan to move away from CPaaS soon and are unlikely to bump or add dependencies.

Reviewer: Cool that you remember!

In fact, what I wrote isn't correct in a practical sense: we could install `requirements.txt`, download anything from internet and do whatever else in `render_templates` (as well as in `pre-build-script`). But doing so violates the spirit of CPaaS&Co as a compliant and trusted build system.

These `render_templates`/`pre-build-script` unlike the Dockerfile builds which run in OSBS are executed in CPaaS Jenkins and aren't network-restricted. I think it's kind of intentional because some products need to do funky stuff for builds (like we download blobs) but at the same time these `render_templates`/`pre-build-script` aren't as hardened as Dockerfile builds in OSBS which can create supply chain security problems if we go and do what we want.

Nobody outside the team asked us to abandon `requirements.txt`, rather it was my idea to do that since we know that supply chain attacks on package repos happen regularly.

Author: I feel like this is resolved by https://github.com/stackrox/stackrox/pull/13209/files#diff-cc4b2f05c18ee7b629ad0bd4633ff59a26ffc67ad8607e97e3155c92f4b848eaR19 and will therefore resolve this comment.

# Bundle Helpers

## Python Dependency Management

For hermetic builds with Konflux, we need to provide the full list of resolved dependencies in `requirements.txt`.
These will be prefetched with Cachi2 and made available to the container image build.

> We use Poetry to manage Python dependencies.

Reviewer: ```suggestion
We use Poetry to manage Python dependencies and for generating the complete `requirements.txt`.
```

* to add a new dependency: `poetry add PyYAML==6.0`
* to update existing dependencies: `poetry update`

> While we still build downstream images with CPaaS, you will need to add/update the dependency also in the `render_templates` script.

Reviewer: ```suggestion
For as long as we build downstream images with CPaaS, you should not add or use dependencies that aren't available OOTB in CPaaS in `render_templates` context.
```
